# Capstone Final Project

### Creating the Dataframe

In [1]:
# Importing Pandas and NumPy and preventing a warning message
import pandas as pd
import numpy as np
pd.options.mode.chained_assignment = None

In [2]:
# Scraping the table from the website
url = "https://en.wikipedia.org/wiki/List_of_London_boroughs"
dfs = pd.read_html(url)

print(len(dfs))

5


In [3]:
# Converting the table into a Dataframe
df = dfs[0]
df

Borough     Inner Status  \
0     Barking and Dagenham [note 1]       NaN    NaN   
1                            Barnet       NaN    NaN   
2                            Bexley       NaN    NaN   
3                             Brent       NaN    NaN   
4                           Bromley       NaN    NaN   
5                            Camden       NaN    NaN   
6                           Croydon       NaN    NaN   
7                            Ealing       NaN    NaN   
8                           Enfield       NaN    NaN   
9                Greenwich [note 2]  [note 3]  Royal   
10                          Hackney       NaN    NaN   
11  Hammersmith and Fulham [note 4]       NaN    NaN   
12                         Haringey  [note 3]    NaN   
13                           Harrow       NaN    NaN   
14                         Havering       NaN    NaN   
15                       Hillingdon       NaN    NaN   
16                         Hounslow       NaN    NaN   
17                        Islington       NaN    NaN   
18           Kensington and Chelsea       NaN  Royal   
19             Kingston upon Thames       NaN  Royal   
20                          Lambeth       NaN    NaN   
21                         Lewisham       NaN    NaN   
22                           Merton       NaN    NaN   
23                           Newham  [note 3]    NaN   
24                        Redbridge       NaN    NaN   
25             Richmond upon Thames       NaN    NaN   
26                        Southwark       NaN    NaN   
27                           Sutton       NaN    NaN   
28                    Tower Hamlets       NaN    NaN   
29                   Waltham Forest       NaN    NaN   
30                       Wandsworth       NaN    NaN   
31                      Westminster       NaN   City   

                                  Local authority           Political control  \
0     Barking and Dagenham London Borough Council                      Labour   
1                   Barnet London Borough Council                Conservative   
2                   Bexley London Borough Council                Conservative   
3                    Brent London Borough Council                      Labour   
4                  Bromley London Borough Council                Conservative   
5                   Camden London Borough Council                      Labour   
6                  Croydon London Borough Council                      Labour   
7                   Ealing London Borough Council                      Labour   
8                  Enfield London Borough Council                      Labour   
9                Greenwich London Borough Council                      Labour   
10                 Hackney London Borough Council                      Labour   
11  Hammersmith and Fulham London Borough Council                      Labour   
12                Haringey London Borough Council                      Labour   
13                  Harrow London Borough Council                      Labour   
14                Havering London Borough Council  Conservative (council NOC)   
15              Hillingdon London Borough Council                Conservative   
16                Hounslow London Borough Council                      Labour   
17               Islington London Borough Council                      Labour   
18  Kensington and Chelsea London Borough Council                Conservative   
19    Kingston upon Thames London Borough Council            Liberal Democrat   
20                 Lambeth London Borough Council                      Labour   
21                Lewisham London Borough Council                      Labour   
22                  Merton London Borough Council                      Labour   
23                  Newham London Borough Council                      Labour   
24               Redbridge London Borough Council                      Labour   
25    Richmond upon Thames London Borough Council            Liberal Demo

### Cleaning the Dataframe

In [4]:
# Only taking the columns of interest
df = df[["Borough", "Co-ordinates"]]
df

Borough  \
0     Barking and Dagenham [note 1]   
1                            Barnet   
2                            Bexley   
3                             Brent   
4                           Bromley   
5                            Camden   
6                           Croydon   
7                            Ealing   
8                           Enfield   
9                Greenwich [note 2]   
10                          Hackney   
11  Hammersmith and Fulham [note 4]   
12                         Haringey   
13                           Harrow   
14                         Havering   
15                       Hillingdon   
16                         Hounslow   
17                        Islington   
18           Kensington and Chelsea   
19             Kingston upon Thames   
20                          Lambeth   
21                         Lewisham   
22                           Merton   
23                           Newham   
24                        Redbridge   
25             Richmond upon Thames   
26                        Southwark   
27                           Sutton   
28                    Tower Hamlets   
29                   Waltham Forest   
30                       Wandsworth   
31                      Westminster   

                                         Co-ordinates  
0   .mw-parser-output .geo-default,.mw-parser-outp...  
1         51°37′31″N 0°09′06″W / 51.6252°N 0.1517°W  
2         51°27′18″N 0°09′02″E / 51.4549°N 0.1505°E  
3         51°33′32″N 0°16′54″W / 51.5588°N 0.2817°W  
4         51°24′14″N 0°01′11″E / 51.4039°N 0.0198°E  
5         51°31′44″N 0°07′32″W / 51.5290°N 0.1255°W  
6         51°22′17″N 0°05′52″W / 51.3714°N 0.0977°W  
7         51°30′47″N 0°18′32″W / 51.5130°N 0.3089°W  
8         51°39′14″N 0°04′48″W / 51.6538°N 0.0799°W  
9         51°29′21″N 0°03′53″E / 51.4892°N 0.0648°E  
10        51°32′42″N 0°03′19″W / 51.5450°N 0.0553°W  
11        51°29′34″N 0°14′02″W / 51.4927°N 0.2339°W  
12        51°36′00″N 0°06′43″W / 51.6000°N 0.1119°W  
13        51°35′23″N 0°20′05″W / 51.5898°N 0.3346°W  
14        51°34′52″N 0°11′01″E / 51.5812°N 0.1837°E  
15        51°32′39″N 0°28′34″W / 51.5441°N 0.4760°W  
16        51°28′29″N 0°22′05″W / 51.4746°N 0.3680°W  
17        51°32′30″N 0°06′08″W / 51.5416°N 0.1022°W  
18        51°30′07″N 0°11′41″W / 51.5020°N 0.1947°W  
19        51°24′31″N 0°18′23″W / 51.4085°N 0.3064°W  
20        51°27′39″N 0°06′59″W / 51.4607°N 0.1163°W  
21        51°26′43″N 0°01′15″W / 51.4452°N 0.0209°W  
22        51°24′05″N 0°11′45″W / 51.4014°N 0.1958°W  
23        51°30′28″N 0°02′49″E / 51.5077°N 0.0469°E  
24        51°33′32″N 0°04′27″E / 51.5590°N 0.0741°E  
25        51°26′52″N 0°19′34″W / 51.4479°N 0.3260°W  
26        51°30′13″N 0°04′49″W / 51.5035°N 0.0804°W  
27        51°21′42″N 0°11′40″W / 51.3618°N 0.1945°W  
28        51°30′36″N 0°00′21″W / 51.5099°N 0.0059°W  
29        51°35′27″N 0°00′48″W / 51.5908°N 0.0134°W  
30        51°27′24″N 0°11′28″W / 51.4567°N 0.1910°W  
31        51°29′50″N 0°08′14″W / 51.4973°N 0.1372°W

In [5]:
# Dropping the first row as there are no coordinate values
df.drop([0], inplace = True)
df.reset_index(drop = True, inplace = True)
df

Borough  \
0                            Barnet   
1                            Bexley   
2                             Brent   
3                           Bromley   
4                            Camden   
5                           Croydon   
6                            Ealing   
7                           Enfield   
8                Greenwich [note 2]   
9                           Hackney   
10  Hammersmith and Fulham [note 4]   
11                         Haringey   
12                           Harrow   
13                         Havering   
14                       Hillingdon   
15                         Hounslow   
16                        Islington   
17           Kensington and Chelsea   
18             Kingston upon Thames   
19                          Lambeth   
20                         Lewisham   
21                           Merton   
22                           Newham   
23                        Redbridge   
24             Richmond upon Thames   
25                        Southwark   
26                           Sutton   
27                    Tower Hamlets   
28                   Waltham Forest   
29                       Wandsworth   
30                      Westminster   

                                   Co-ordinates  
0   51°37′31″N 0°09′06″W / 51.6252°N 0.1517°W  
1   51°27′18″N 0°09′02″E / 51.4549°N 0.1505°E  
2   51°33′32″N 0°16′54″W / 51.5588°N 0.2817°W  
3   51°24′14″N 0°01′11″E / 51.4039°N 0.0198°E  
4   51°31′44″N 0°07′32″W / 51.5290°N 0.1255°W  
5   51°22′17″N 0°05′52″W / 51.3714°N 0.0977°W  
6   51°30′47″N 0°18′32″W / 51.5130°N 0.3089°W  
7   51°39′14″N 0°04′48″W / 51.6538°N 0.0799°W  
8   51°29′21″N 0°03′53″E / 51.4892°N 0.0648°E  
9   51°32′42″N 0°03′19″W / 51.5450°N 0.0553°W  
10  51°29′34″N 0°14′02″W / 51.4927°N 0.2339°W  
11  51°36′00″N 0°06′43″W / 51.6000°N 0.1119°W  
12  51°35′23″N 0°20′05″W / 51.5898°N 0.3346°W  
13  51°34′52″N 0°11′01″E / 51.5812°N 0.1837°E  
14  51°32′39″N 0°28′34″W / 51.5441°N 0.4760°W  
15  51°28′29″N 0°22′05″W / 51.4746°N 0.3680°W  
16  51°32′30″N 0°06′08″W / 51.5416°N 0.1022°W  
17  51°30′07″N 0°11′41″W / 51.5020°N 0.1947°W  
18  51°24′31″N 0°18′23″W / 51.4085°N 0.3064°W  
19  51°27′39″N 0°06′59″W / 51.4607°N 0.1163°W  
20  51°26′43″N 0°01′15″W / 51.4452°N 0.0209°W  
21  51°24′05″N 0°11′45″W / 51.4014°N 0.1958°W  
22  51°30′28″N 0°02′49″E / 51.5077°N 0.0469°E  
23  51°33′32″N 0°04′27″E / 51.5590°N 0.0741°E  
24  51°26′52″N 0°19′34″W / 51.4479°N 0.3260°W  
25  51°30′13″N 0°04′49″W / 51.5035°N 0.0804°W  
26  51°21′42″N 0°11′40″W / 51.3618°N 0.1945°W  
27  51°30′36″N 0°00′21″W / 51.5099°N 0.0059°W  
28  51°35′27″N 0°00′48″W / 51.5908°N 0.0134°W  
29  51°27′24″N 0°11′28″W / 51.4567°N 0.1910°W  
30  51°29′50″N 0°08′14″W / 51.4973°N 0.1372°W

In [6]:
# Removing the unnecessary 'notes' in the Borough names
df.replace({"Borough": {"Greenwich [note 2]": "Greenwich", 
                        "Hammersmith and Fulham [note 4]": "Hammersmith and Fulham"}}, inplace = True)
df

Borough                                 Co-ordinates
0                   Barnet  51°37′31″N 0°09′06″W / 51.6252°N 0.1517°W
1                   Bexley  51°27′18″N 0°09′02″E / 51.4549°N 0.1505°E
2                    Brent  51°33′32″N 0°16′54″W / 51.5588°N 0.2817°W
3                  Bromley  51°24′14″N 0°01′11″E / 51.4039°N 0.0198°E
4                   Camden  51°31′44″N 0°07′32″W / 51.5290°N 0.1255°W
5                  Croydon  51°22′17″N 0°05′52″W / 51.3714°N 0.0977°W
6                   Ealing  51°30′47″N 0°18′32″W / 51.5130°N 0.3089°W
7                  Enfield  51°39′14″N 0°04′48″W / 51.6538°N 0.0799°W
8                Greenwich  51°29′21″N 0°03′53″E / 51.4892°N 0.0648°E
9                  Hackney  51°32′42″N 0°03′19″W / 51.5450°N 0.0553°W
10  Hammersmith and Fulham  51°29′34″N 0°14′02″W / 51.4927°N 0.2339°W
11                Haringey  51°36′00″N 0°06′43″W / 51.6000°N 0.1119°W
12                  Harrow  51°35′23″N 0°20′05″W / 51.5898°N 0.3346°W
13                Havering  51°34′52″N 0°11′01″E / 51.5812°N 0.1837°E
14              Hillingdon  51°32′39″N 0°28′34″W / 51.5441°N 0.4760°W
15                Hounslow  51°28′29″N 0°22′05″W / 51.4746°N 0.3680°W
16               Islington  51°32′30″N 0°06′08″W / 51.5416°N 0.1022°W
17  Kensington and Chelsea  51°30′07″N 0°11′41″W / 51.5020°N 0.1947°W
18    Kingston upon Thames  51°24′31″N 0°18′23″W / 51.4085°N 0.3064°W
19                 Lambeth  51°27′39″N 0°06′59″W / 51.4607°N 0.1163°W
20                Lewisham  51°26′43″N 0°01′15″W / 51.4452°N 0.0209°W
21                  Merton  51°24′05″N 0°11′45″W / 51.4014°N 0.1958°W
22                  Newham  51°30′28″N 0°02′49″E / 51.5077°N 0.0469°E
23               Redbridge  51°33′32″N 0°04′27″E / 51.5590°N 0.0741°E
24    Richmond upon Thames  51°26′52″N 0°19′34″W / 51.4479°N 0.3260°W
25               Southwark  51°30′13″N 0°04′49″W / 51.5035°N 0.0804°W
26                  Sutton  51°21′42″N 0°11′40″W / 51.3618°N 0.1945°W
27           Tower Hamlets  51°30′36″N 0°00′21″W / 51.5099°N 0.0059°W
28          Waltham Forest  51°35′27″N 0°00′48″W / 51.5908°N 0.0134°W
29              Wandsworth  51°27′24″N 0°11′28″W / 51.4567°N 0.1910°W
30             Westminster  51°29′50″N 0°08′14″W / 51.4973°N 0.1372°W

In [7]:
# Splitting the coordinates column to only have the needed values
df[["LatLong1", "LatLong2"]] = df["Co-ordinates"].str.split("/", expand = True)
df = df[["Borough", "LatLong2"]]
df

Borough              LatLong2
0                   Barnet   51.6252°N 0.1517°W
1                   Bexley   51.4549°N 0.1505°E
2                    Brent   51.5588°N 0.2817°W
3                  Bromley   51.4039°N 0.0198°E
4                   Camden   51.5290°N 0.1255°W
5                  Croydon   51.3714°N 0.0977°W
6                   Ealing   51.5130°N 0.3089°W
7                  Enfield   51.6538°N 0.0799°W
8                Greenwich   51.4892°N 0.0648°E
9                  Hackney   51.5450°N 0.0553°W
10  Hammersmith and Fulham   51.4927°N 0.2339°W
11                Haringey   51.6000°N 0.1119°W
12                  Harrow   51.5898°N 0.3346°W
13                Havering   51.5812°N 0.1837°E
14              Hillingdon   51.5441°N 0.4760°W
15                Hounslow   51.4746°N 0.3680°W
16               Islington   51.5416°N 0.1022°W
17  Kensington and Chelsea   51.5020°N 0.1947°W
18    Kingston upon Thames   51.4085°N 0.3064°W
19                 Lambeth   51.4607°N 0.1163°W
20                Lewisham   51.4452°N 0.0209°W
21                  Merton   51.4014°N 0.1958°W
22                  Newham   51.5077°N 0.0469°E
23               Redbridge   51.5590°N 0.0741°E
24    Richmond upon Thames   51.4479°N 0.3260°W
25               Southwark   51.5035°N 0.0804°W
26                  Sutton   51.3618°N 0.1945°W
27           Tower Hamlets   51.5099°N 0.0059°W
28          Waltham Forest   51.5908°N 0.0134°W
29              Wandsworth   51.4567°N 0.1910°W
30             Westminster   51.4973°N 0.1372°W

In [8]:
# Splitting the LatLong2 column into separate latitude and longitude columns
df1 = df["LatLong2"].str.split(" ", expand = True)
df1.columns = ["LongLat{}".format(x) for x in df1.columns]
df = df.join(df1)
df = df[["Borough", "LongLat1", "LongLat2"]]

# Renaming the columns
df.rename(columns = {"LongLat1": "Latitude", 
                     "LongLat2": "Longitude"}, inplace  = True)
df

Borough    Latitude Longitude
0                   Barnet  51.6252°N  0.1517°W
1                   Bexley  51.4549°N  0.1505°E
2                    Brent  51.5588°N  0.2817°W
3                  Bromley  51.4039°N  0.0198°E
4                   Camden  51.5290°N  0.1255°W
5                  Croydon  51.3714°N  0.0977°W
6                   Ealing  51.5130°N  0.3089°W
7                  Enfield  51.6538°N  0.0799°W
8                Greenwich  51.4892°N  0.0648°E
9                  Hackney  51.5450°N  0.0553°W
10  Hammersmith and Fulham  51.4927°N  0.2339°W
11                Haringey  51.6000°N  0.1119°W
12                  Harrow  51.5898°N  0.3346°W
13                Havering  51.5812°N  0.1837°E
14              Hillingdon  51.5441°N  0.4760°W
15                Hounslow  51.4746°N  0.3680°W
16               Islington  51.5416°N  0.1022°W
17  Kensington and Chelsea  51.5020°N  0.1947°W
18    Kingston upon Thames  51.4085°N  0.3064°W
19                 Lambeth  51.4607°N  0.1163°W
20                Lewisham  51.4452°N  0.0209°W
21                  Merton  51.4014°N  0.1958°W
22                  Newham  51.5077°N  0.0469°E
23               Redbridge  51.5590°N  0.0741°E
24    Richmond upon Thames  51.4479°N  0.3260°W
25               Southwark  51.5035°N  0.0804°W
26                  Sutton  51.3618°N  0.1945°W
27           Tower Hamlets  51.5099°N  0.0059°W
28          Waltham Forest  51.5908°N  0.0134°W
29              Wandsworth  51.4567°N  0.1910°W
30             Westminster  51.4973°N  0.1372°W

In [9]:
# Removing the unnecessary '°N' from the latitude values
df["Latitude"] = df["Latitude"].str.rstrip("°N")
df

Borough  Latitude Longitude
0                   Barnet  51.6252  0.1517°W
1                   Bexley  51.4549  0.1505°E
2                    Brent  51.5588  0.2817°W
3                  Bromley  51.4039  0.0198°E
4                   Camden  51.5290  0.1255°W
5                  Croydon  51.3714  0.0977°W
6                   Ealing  51.5130  0.3089°W
7                  Enfield  51.6538  0.0799°W
8                Greenwich  51.4892  0.0648°E
9                  Hackney  51.5450  0.0553°W
10  Hammersmith and Fulham  51.4927  0.2339°W
11                Haringey  51.6000  0.1119°W
12                  Harrow  51.5898  0.3346°W
13                Havering  51.5812  0.1837°E
14              Hillingdon  51.5441  0.4760°W
15                Hounslow  51.4746  0.3680°W
16               Islington  51.5416  0.1022°W
17  Kensington and Chelsea  51.5020  0.1947°W
18    Kingston upon Thames  51.4085  0.3064°W
19                 Lambeth  51.4607  0.1163°W
20                Lewisham  51.4452  0.0209°W
21                  Merton  51.4014  0.1958°W
22                  Newham  51.5077  0.0469°E
23               Redbridge  51.5590  0.0741°E
24    Richmond upon Thames  51.4479  0.3260°W
25               Southwark  51.5035  0.0804°W
26                  Sutton  51.3618  0.1945°W
27           Tower Hamlets  51.5099  0.0059°W
28          Waltham Forest  51.5908  0.0134°W
29              Wandsworth  51.4567  0.1910°W
30             Westminster  51.4973  0.1372°W

In [10]:
# Adjusting the longitude values so they are in the correct format
london_df = df

london_df["Longitude"] = london_df["Longitude"].str.rstrip("°W")
london_df["Longitude"] = np.where(london_df["Longitude"].str.len() == 6, 
                                  "-" + london_df["Longitude"], 
                                  london_df["Longitude"].str.rstrip("°E"))

london_df

Borough  Latitude Longitude
0                   Barnet  51.6252   -0.1517
1                   Bexley  51.4549    0.1505
2                    Brent  51.5588   -0.2817
3                  Bromley  51.4039    0.0198
4                   Camden  51.5290   -0.1255
5                  Croydon  51.3714   -0.0977
6                   Ealing  51.5130   -0.3089
7                  Enfield  51.6538   -0.0799
8                Greenwich  51.4892    0.0648
9                  Hackney  51.5450   -0.0553
10  Hammersmith and Fulham  51.4927   -0.2339
11                Haringey  51.6000   -0.1119
12                  Harrow  51.5898   -0.3346
13                Havering  51.5812    0.1837
14              Hillingdon  51.5441   -0.4760
15                Hounslow  51.4746   -0.3680
16               Islington  51.5416   -0.1022
17  Kensington and Chelsea  51.5020   -0.1947
18    Kingston upon Thames  51.4085   -0.3064
19                 Lambeth  51.4607   -0.1163
20                Lewisham  51.4452   -0.0209
21                  Merton  51.4014   -0.1958
22                  Newham  51.5077    0.0469
23               Redbridge  51.5590    0.0741
24    Richmond upon Thames  51.4479   -0.3260
25               Southwark  51.5035   -0.0804
26                  Sutton  51.3618   -0.1945
27           Tower Hamlets  51.5099   -0.0059
28          Waltham Forest  51.5908   -0.0134
29              Wandsworth  51.4567   -0.1910
30             Westminster  51.4973   -0.1372

In [11]:
# Removing hidden byte order mark from web scraping
london_df["Latitude"] = london_df["Latitude"].str.replace(u"\ufeff", "")
london_df["Longitude"] = london_df["Longitude"].str.replace(u"\ufeff", "")

### Creating a map to visualise the Boroughs

In [12]:
# Importing the necessary programs to make the map
from geopy.geocoders import Nominatim
import folium
import requests

In [13]:
# Getting the latitude and longitude values of London
geolocator = Nominatim(user_agent = "london_explorer")
location = geolocator.geocode("London, UK")
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of London are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of London are 51.5073219, -0.1276474.


In [14]:
#Producing a map of London using the latitude and longitude values
map_london = folium.Map(location = [latitude, longitude], zoom_start = 10)

# Marking all the Boroughs on the map
for lat, lng, borough in zip(london_df["Latitude"], london_df["Longitude"], london_df["Borough"]):
    label = "{}".format(borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker([lat, lng], 
                        radius = 5, 
                        popup = label, 
                        color = "blue", 
                        fill = True, 
                        fill_color = "3186cc", 
                        fill_opacity = 0.7, 
                        parse_html = False).add_to(map_london)

map_london

In [15]:
# This cell contained Foursquare ID and secret

### Using the Foursquare data to get information on venues within the Boroughs

In [16]:
# Creating a function to get venue information for all the Boroughs
def getNearbyVenues(boroughs, latitudes, longitudes, radius = 1000):
    
    venues_list = []
    for borough, lat, lng in zip(boroughs, latitudes, longitudes):
        print(borough)
        
        url = "https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        
        results = requests.get(url).json()["response"]["groups"][0]["items"]
        
        venues_list.append([(
            borough, 
            lat, 
            lng, 
            v["venue"]["name"], 
            v["venue"]["location"]["lat"], 
            v["venue"]["location"]["lng"], 
            v["venue"]["categories"][0]["name"]) for v in results])
        
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ["Borough", 
                             "Borough Latitude", 
                             "Borough Longitude", 
                             "Venue", 
                             "Venue Latitude", 
                             "Venue Longitude", 
                             "Venue Category"]
    
    return(nearby_venues)

In [17]:
# Creating a new Dataframe with this information
london_venues = getNearbyVenues(boroughs = london_df["Borough"], 
                                latitudes = london_df["Latitude"], 
                                longitudes = london_df["Longitude"])
london_venues

Barnet
Bexley
Brent
Bromley
Camden
Croydon
Ealing
Enfield
Greenwich
Hackney
Hammersmith and Fulham
Haringey
Harrow
Havering
Hillingdon
Hounslow
Islington
Kensington and Chelsea
Kingston upon Thames
Lambeth
Lewisham
Merton
Newham
Redbridge
Richmond upon Thames
Southwark
Sutton
Tower Hamlets
Waltham Forest
Wandsworth
Westminster


Borough Borough Latitude Borough Longitude  \
0          Barnet          51.6252           -0.1517   
1          Barnet          51.6252           -0.1517   
2          Barnet          51.6252           -0.1517   
3          Barnet          51.6252           -0.1517   
4          Barnet          51.6252           -0.1517   
...           ...              ...               ...   
2188  Westminster          51.4973           -0.1372   
2189  Westminster          51.4973           -0.1372   
2190  Westminster          51.4973           -0.1372   
2191  Westminster          51.4973           -0.1372   
2192  Westminster          51.4973           -0.1372   

                         Venue  Venue Latitude  Venue Longitude  \
0                 The Cavalier       51.629646        -0.157798   
1                   The Atrium       51.624726        -0.151933   
2       Beaconsfield Road (BF)       51.622827        -0.151466   
3              The Osidge Arms       51.631986        -0.143123   
4                Tesco Express       51.625944        -0.160032   
...                        ...             ...              ...   
2188           The Queens Arms       51.492567        -0.139297   
2189   Upper Grosvenor Gardens       51.497802        -0.147046   
2190      Grays of Westminster       51.491336        -0.138983   
2191  Winston Churchill Statue       51.500799        -0.126267   
2192          Rosa's Thai Cafe       51.493732        -0.141673   

         Venue Category  
0                   Pub  
1                  Café  
2              Bus Stop  
3                   Pub  
4         Grocery Store  
...                 ...  
2188                Pub  
2189               Park  
2190       Camera Store  
2191  Outdoor Sculpture  
2192    Thai Restaurant  

[2193 rows x 7 columns]

In [18]:
# The number of unique types of venues
print("There are {} unique categories".format(len(london_venues["Venue Category"].unique())))

# The number of venues per Borough
london_venues.groupby("Borough").count()

There are 258 unique categories


Borough Latitude  Borough Longitude  Venue  \
Borough                                                              
Barnet                                 8                  8      8   
Bexley                                47                 47     47   
Brent                                 90                 90     90   
Bromley                               62                 62     62   
Camden                               100                100    100   
Croydon                               96                 96     96   
Ealing                               100                100    100   
Enfield                               63                 63     63   
Greenwich                             58                 58     58   
Hackney                               79                 79     79   
Hammersmith and Fulham               100                100    100   
Haringey                              68                 68     68   
Harrow                                42                 42     42   
Havering                              76                 76     76   
Hillingdon                            72                 72     72   
Hounslow                              66                 66     66   
Islington                            100                100    100   
Kensington and Chelsea               100                100    100   
Kingston upon Thames                 100                100    100   
Lambeth                              100                100    100   
Lewisham                              39                 39     39   
Merton                                29                 29     29   
Newham                                38                 38     38   
Redbridge                             45                 45     45   
Richmond upon Thames                  77                 77     77   
Southwark                            100                100    100   
Sutton                                51                 51     51   
Tower Hamlets                         73                 73     73   
Waltham Forest                        39                 39     39   
Wandsworth                            75                 75     75   
Westminster                          100                100    100   

                        Venue Latitude  Venue Longitude  Venue Category  
Borough                                                                  
Barnet                               8                8               8  
Bexley                              47               47              47  
Brent                               90               90              90  
Bromley                             62               62              62  
Camden                             100              100             100  
Croydon                             96               96              96  
Ealing                             100              100             100  
Enfield                             63               63              63  
Greenwich                           58               58              58  
Hackney                             79               79              79  
Hammersmith and Fulham             100              100             100  
Haringey                            68               68              68  
Harrow                              42               42              42  
Havering                            76               76              76  
Hillingdon                          72               72              72  
Hounslow                            66               66              66  
Islington                          100              100             100  
Kensington and Chelsea             100              100             100  
Kingston upon Thames               100              100             100  
Lambeth                            100              100             100  
Lewisham                            39               39              39  
Merton     

### Analysing the Boroughs

The countries defined as 'East Asian' (China, Hong Kong, Japan, Macau, Mongolia, North Korea, South Korea and Taiwan) would be expected to have similar cuisines to that in South Korea.

In [19]:
# Getting all the unique venue categories to separate out the 'East Asian' restaurants
print(london_venues["Venue Category"].unique())

['Pub' 'Café' 'Bus Stop' 'Grocery Store' 'Park' 'Fish & Chips Shop'
 'Italian Restaurant' 'Furniture / Home Store' 'Clothing Store'
 'American Restaurant' 'Hotel' 'Plaza' 'Pharmacy' 'Department Store'
 'Supermarket' 'Bakery' 'Coffee Shop' 'Video Game Store'
 'Portuguese Restaurant' 'Pizza Place' 'Fast Food Restaurant'
 'Chinese Restaurant' 'Warehouse Store' 'Shopping Mall' 'Restaurant'
 'Mexican Restaurant' 'Bus Station' 'Burger Joint'
 'Mediterranean Restaurant' 'Nightclub' 'Greek Restaurant' 'Playground'
 'Bowling Alley' 'Stadium' 'Chocolate Shop' 'Outdoor Sculpture'
 'Food Court' 'Music Venue' 'Multiplex' 'Sandwich Place'
 'Latin American Restaurant' 'Sporting Goods Shop' 'Bar' 'Bistro'
 'Roof Deck' 'Indian Restaurant' 'Ice Cream Shop' 'Gym / Fitness Center'
 'Asian Restaurant' 'Outlet Mall' 'Hotel Bar' 'Post Office'
 'Bubble Tea Shop' 'Pedestrian Plaza' 'Sports Bar' 'Music Store'
 'Outlet Store' 'Electronics Store' 'Train Station' 'Diner'
 'Kitchen Supply Store' 'Seafood Restaurant

In [20]:
# Separating out the 'East Asian' types of restaurants
east_asian_cuisine = ["Chinese Restaurant", "Asian Restaurant", "Sushi Restaurant", "Korean Restaurant", 
                      "Noodle House", "Ramen Restaurant", "Taiwanese Restaurant", "Japanese Restaurant", 
                      "Hunan Restaurant", "Okonomiyaki Restaurant", "Dumpling Restaurant", "Dim Sum Restaurant", 
                      "Cantonese Restaurant"]

In [21]:
# One hot encoding each venue type for each Borough
london_onehot = pd.get_dummies(london_venues[["Venue Category"]], prefix = "", prefix_sep = "")

# Adding the Borough column to the new Dataframe
london_onehot["Borough"] = london_venues["Borough"]

# Re-ordering the columns to make sure the Borough column is first in the Dataframe
fixed_columns = [london_onehot.columns[-1]] + list(london_onehot.columns[:-1])
london_onehot = london_onehot[fixed_columns]

london_onehot

Borough  Afghan Restaurant  African Restaurant  Airport  \
0          Barnet                  0                   0        0   
1          Barnet                  0                   0        0   
2          Barnet                  0                   0        0   
3          Barnet                  0                   0        0   
4          Barnet                  0                   0        0   
...           ...                ...                 ...      ...   
2188  Westminster                  0                   0        0   
2189  Westminster                  0                   0        0   
2190  Westminster                  0                   0        0   
2191  Westminster                  0                   0        0   
2192  Westminster                  0                   0        0   

      Airport Lounge  Airport Service  Airport Ticket Counter  \
0                  0                0                       0   
1                  0                0                       0   
2                  0                0                       0   
3                  0                0                       0   
4                  0                0                       0   
...              ...              ...                     ...   
2188               0                0                       0   
2189               0                0                       0   
2190               0                0                       0   
2191               0                0                       0   
2192               0                0                       0   

      American Restaurant  Argentinian Restaurant  Art Gallery  ...  \
0                       0                       0            0  ...   
1                       0                       0            0  ...   
2                       0                       0            0  ...   
3                       0                       0            0  ...   
4                       0                       0            0  ...   
...                   ...                     ...          ...  ...   
2188                    0                       0            0  ...   
2189                    0                       0            0  ...   
2190                    0                       0            0  ...   
2191                    0                       0            0  ...   
2192                    0                       0            0  ...   

      Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                      0                              0                 0   
1                      0                              0                 0   
2                      0                              0                 0   
3                      0                              0                 0   
4                      0                              0                 0   
...                  ...                            ...               ...   
2188                   0                              0                 0   
2189                   0                              0                 0   
2190                   0                              0                 0   
2191                   0                              0                 0   
2192                   0                              0                 0   

      Vietnamese Restaurant  Warehouse Store  Windmill  Wine Bar  Wine Shop  \
0                         0                0         0         0          0   
1                         0                0         0         0          0   
2                         0                0         0         0          0   
3                         0                0         0         0          0   
4                         0                0         0         0          0   
...                     ...              ...       ...       ...        ...   
2188                      0                0         0         0     

In [22]:
# Grouping the rows by Borough and taking the mean of the frequency of each category
london_grouped = london_onehot.groupby(["Borough"]).mean().reset_index()
london_grouped

Borough  Afghan Restaurant  African Restaurant   Airport  \
0                   Barnet            0.00000            0.000000  0.000000   
1                   Bexley            0.00000            0.000000  0.000000   
2                    Brent            0.00000            0.000000  0.000000   
3                  Bromley            0.00000            0.000000  0.000000   
4                   Camden            0.00000            0.000000  0.000000   
5                  Croydon            0.00000            0.000000  0.000000   
6                   Ealing            0.00000            0.000000  0.000000   
7                  Enfield            0.00000            0.000000  0.000000   
8                Greenwich            0.00000            0.017241  0.000000   
9                  Hackney            0.00000            0.000000  0.000000   
10  Hammersmith and Fulham            0.00000            0.000000  0.000000   
11                Haringey            0.00000            0.000000  0.000000   
12                  Harrow            0.02381            0.000000  0.000000   
13                Havering            0.00000            0.000000  0.000000   
14              Hillingdon            0.00000            0.000000  0.000000   
15                Hounslow            0.00000            0.000000  0.000000   
16               Islington            0.01000            0.000000  0.000000   
17  Kensington and Chelsea            0.00000            0.000000  0.000000   
18    Kingston upon Thames            0.00000            0.000000  0.000000   
19                 Lambeth            0.00000            0.020000  0.000000   
20                Lewisham            0.00000            0.000000  0.000000   
21                  Merton            0.00000            0.000000  0.000000   
22                  Newham            0.00000            0.000000  0.026316   
23               Redbridge            0.00000            0.000000  0.000000   
24    Richmond upon Thames            0.00000            0.000000  0.000000   
25               Southwark            0.00000            0.000000  0.000000   
26                  Sutton            0.00000            0.000000  0.000000   
27           Tower Hamlets            0.00000            0.000000  0.000000   
28          Waltham Forest            0.00000            0.000000  0.000000   
29              Wandsworth            0.00000            0.000000  0.000000   
30             Westminster            0.00000            0.000000  0.000000   

    Airport Lounge  Airport Service  Airport Ticket Counter  \
0         0.000000         0.000000                0.000000   
1         0.000000         0.000000                0.000000   
2         0.000000         0.000000                0.000000   
3         0.000000         0.000000                0.000000   
4         0.000000         0.000000                0.000000   
5         0.000000         0.000000                0.000000   
6         0.000000         0.000000                0.000000   
7         0.000000         0.000000                0.000000   
8         0.000000         0.000000                0.000000   
9         0.000000         0.000000                0.000000   
10        0.000000         0.000000                0.000000   
11        0.000000         0.000000                0.000000   
12        0.000000         0.000000                0.000000   
13        0.000000         0.000000                0.000000   
14        0.000000         0.000000                0.000000   
15        0.000000         0.000000                0.000000   
16        0.000000         0.000000                0.000000   
17        0.000000         0.000000                0.000000   
18        0.000000         0.000000                0.000000   
19        0.000000         0.000000                0.000000   
20        0.000000         0.000000                0.000000   
21        0.000000         0.000000                0.000000   
22        0.026316         0.026316      

In [23]:
# Creating a new Dataframe containing just the 'East Asian' restaurants in each Borough
london_asian_restaurants = london_grouped[["Borough", "Asian Restaurant", "Cantonese Restaurant", "Chinese Restaurant", 
                                           "Dim Sum Restaurant", "Dumpling Restaurant", "Hunan Restaurant", 
                                           "Japanese Restaurant", "Korean Restaurant", "Noodle House", "Okonomiyaki Restaurant",
                                           "Ramen Restaurant", "Sushi Restaurant", "Taiwanese Restaurant"]]
london_asian_restaurants

Borough  Asian Restaurant  Cantonese Restaurant  \
0                   Barnet          0.000000                  0.00   
1                   Bexley          0.000000                  0.00   
2                    Brent          0.011111                  0.00   
3                  Bromley          0.016129                  0.00   
4                   Camden          0.010000                  0.00   
5                  Croydon          0.020833                  0.00   
6                   Ealing          0.010000                  0.00   
7                  Enfield          0.000000                  0.00   
8                Greenwich          0.017241                  0.00   
9                  Hackney          0.000000                  0.00   
10  Hammersmith and Fulham          0.010000                  0.00   
11                Haringey          0.000000                  0.00   
12                  Harrow          0.000000                  0.00   
13                Havering          0.000000                  0.00   
14              Hillingdon          0.013889                  0.00   
15                Hounslow          0.015152                  0.00   
16               Islington          0.000000                  0.00   
17  Kensington and Chelsea          0.010000                  0.00   
18    Kingston upon Thames          0.010000                  0.00   
19                 Lambeth          0.000000                  0.01   
20                Lewisham          0.000000                  0.00   
21                  Merton          0.000000                  0.00   
22                  Newham          0.000000                  0.00   
23               Redbridge          0.000000                  0.00   
24    Richmond upon Thames          0.000000                  0.00   
25               Southwark          0.010000                  0.00   
26                  Sutton          0.000000                  0.00   
27           Tower Hamlets          0.000000                  0.00   
28          Waltham Forest          0.000000                  0.00   
29              Wandsworth          0.026667                  0.00   
30             Westminster          0.000000                  0.00   

    Chinese Restaurant  Dim Sum Restaurant  Dumpling Restaurant  \
0             0.000000                0.00                 0.00   
1             0.021277                0.00                 0.00   
2             0.000000                0.00                 0.00   
3             0.016129                0.00                 0.00   
4             0.000000                0.00                 0.00   
5             0.010417                0.00                 0.00   
6             0.000000                0.00                 0.00   
7             0.000000                0.00                 0.00   
8             0.017241                0.00                 0.00   
9             0.000000                0.00                 0.00   
10            0.000000                0.00                 0.00   
11            0.044118                0.00                 0.00   
12            0.023810                0.00                 0.00   
13            0.000000                0.00                 0.00   
14            0.013889                0.00                 0.00   
15            0.015152                0.00                 0.00   
16            0.000000                0.00                 0.00   
17            0.020000                0.00                 0.00   
18            0.010000                0.00                 0.00   
19            0.000000                0.01                 0.01   
20            0.025641                0.00                 0.00   
21            0.000000                0.00                 0.00   
22            0.026316                0.00                 0.00   
23            0.000000                0.00                 0.00   
24            0.012987                0.00                 0.00   
25            0.000000              

In [24]:
# Summing up all the frequencies in each Borough
summed_restaurants = london_asian_restaurants.sum(axis = 1)
summed_restaurants

0     0.000000
1     0.021277
2     0.011111
3     0.048387
4     0.020000
5     0.062500
6     0.040000
7     0.000000
8     0.034483
9     0.037975
10    0.050000
11    0.044118
12    0.023810
13    0.000000
14    0.027778
15    0.030303
16    0.040000
17    0.050000
18    0.070000
19    0.070000
20    0.051282
21    0.034483
22    0.026316
23    0.000000
24    0.025974
25    0.010000
26    0.000000
27    0.041096
28    0.000000
29    0.026667
30    0.040000
dtype: float64

In [25]:
# Adding the summed values to the main Dataframe
london_asian_restaurants["Total"] = summed_restaurants
london_asian_restaurants

Borough  Asian Restaurant  Cantonese Restaurant  \
0                   Barnet          0.000000                  0.00   
1                   Bexley          0.000000                  0.00   
2                    Brent          0.011111                  0.00   
3                  Bromley          0.016129                  0.00   
4                   Camden          0.010000                  0.00   
5                  Croydon          0.020833                  0.00   
6                   Ealing          0.010000                  0.00   
7                  Enfield          0.000000                  0.00   
8                Greenwich          0.017241                  0.00   
9                  Hackney          0.000000                  0.00   
10  Hammersmith and Fulham          0.010000                  0.00   
11                Haringey          0.000000                  0.00   
12                  Harrow          0.000000                  0.00   
13                Havering          0.000000                  0.00   
14              Hillingdon          0.013889                  0.00   
15                Hounslow          0.015152                  0.00   
16               Islington          0.000000                  0.00   
17  Kensington and Chelsea          0.010000                  0.00   
18    Kingston upon Thames          0.010000                  0.00   
19                 Lambeth          0.000000                  0.01   
20                Lewisham          0.000000                  0.00   
21                  Merton          0.000000                  0.00   
22                  Newham          0.000000                  0.00   
23               Redbridge          0.000000                  0.00   
24    Richmond upon Thames          0.000000                  0.00   
25               Southwark          0.010000                  0.00   
26                  Sutton          0.000000                  0.00   
27           Tower Hamlets          0.000000                  0.00   
28          Waltham Forest          0.000000                  0.00   
29              Wandsworth          0.026667                  0.00   
30             Westminster          0.000000                  0.00   

    Chinese Restaurant  Dim Sum Restaurant  Dumpling Restaurant  \
0             0.000000                0.00                 0.00   
1             0.021277                0.00                 0.00   
2             0.000000                0.00                 0.00   
3             0.016129                0.00                 0.00   
4             0.000000                0.00                 0.00   
5             0.010417                0.00                 0.00   
6             0.000000                0.00                 0.00   
7             0.000000                0.00                 0.00   
8             0.017241                0.00                 0.00   
9             0.000000                0.00                 0.00   
10            0.000000                0.00                 0.00   
11            0.044118                0.00                 0.00   
12            0.023810                0.00                 0.00   
13            0.000000                0.00                 0.00   
14            0.013889                0.00                 0.00   
15            0.015152                0.00                 0.00   
16            0.000000                0.00                 0.00   
17            0.020000                0.00                 0.00   
18            0.010000                0.00                 0.00   
19            0.000000                0.01                 0.01   
20            0.025641                0.00                 0.00   
21            0.000000                0.00                 0.00   
22            0.026316                0.00                 0.00   
23            0.000000                0.00                 0.00   
24            0.012987                0.00                 0.00   
25            0.000000              

In [26]:
# Sorting the Dataframe so the Boroughs with the highest frequency of 'East Asian' restaurants are at the top
sorted_restaurants = london_asian_restaurants.sort_values(by = ["Total"], ascending = False, ignore_index = True)
sorted_restaurants

Borough  Asian Restaurant  Cantonese Restaurant  \
0                  Lambeth          0.000000                  0.01   
1     Kingston upon Thames          0.010000                  0.00   
2                  Croydon          0.020833                  0.00   
3                 Lewisham          0.000000                  0.00   
4   Hammersmith and Fulham          0.010000                  0.00   
5   Kensington and Chelsea          0.010000                  0.00   
6                  Bromley          0.016129                  0.00   
7                 Haringey          0.000000                  0.00   
8            Tower Hamlets          0.000000                  0.00   
9              Westminster          0.000000                  0.00   
10               Islington          0.000000                  0.00   
11                  Ealing          0.010000                  0.00   
12                 Hackney          0.000000                  0.00   
13                  Merton          0.000000                  0.00   
14               Greenwich          0.017241                  0.00   
15                Hounslow          0.015152                  0.00   
16              Hillingdon          0.013889                  0.00   
17              Wandsworth          0.026667                  0.00   
18                  Newham          0.000000                  0.00   
19    Richmond upon Thames          0.000000                  0.00   
20                  Harrow          0.000000                  0.00   
21                  Bexley          0.000000                  0.00   
22                  Camden          0.010000                  0.00   
23                   Brent          0.011111                  0.00   
24               Southwark          0.010000                  0.00   
25                Havering          0.000000                  0.00   
26                 Enfield          0.000000                  0.00   
27               Redbridge          0.000000                  0.00   
28                  Sutton          0.000000                  0.00   
29          Waltham Forest          0.000000                  0.00   
30                  Barnet          0.000000                  0.00   

    Chinese Restaurant  Dim Sum Restaurant  Dumpling Restaurant  \
0             0.000000                0.01                 0.01   
1             0.010000                0.00                 0.00   
2             0.010417                0.00                 0.00   
3             0.025641                0.00                 0.00   
4             0.000000                0.00                 0.00   
5             0.020000                0.00                 0.00   
6             0.016129                0.00                 0.00   
7             0.044118                0.00                 0.00   
8             0.013699                0.00                 0.00   
9             0.010000                0.00                 0.00   
10            0.000000                0.00                 0.00   
11            0.000000                0.00                 0.00   
12            0.000000                0.00                 0.00   
13            0.000000                0.00                 0.00   
14            0.017241                0.00                 0.00   
15            0.015152                0.00                 0.00   
16            0.013889                0.00                 0.00   
17            0.000000                0.00                 0.00   
18            0.026316                0.00                 0.00   
19            0.012987                0.00                 0.00   
20            0.023810                0.00                 0.00   
21            0.021277                0.00                 0.00   
22            0.000000                0.00                 0.00   
23            0.000000                0.00                 0.00   
24            0.000000                0.00                 0.00   
25            0.000000              

This Dataframe shows that the Boroughs with the highest proportion of 'East Asian' restaurants are Lambeth and Kingston upon Thames, both with a frequency of 0.07, followed by Croydon and Lewisham (0.0625 and 0.0513 respectively).

In [27]:
# Sorting the Dataframe so the Boroughs with the highest frequency of Korean restaurants are at the top
sorted_restaurants_korean = london_asian_restaurants.sort_values(by = ["Korean Restaurant"], ascending = False, 
                                                                 ignore_index = True)
sorted_restaurants_korean

Borough  Asian Restaurant  Cantonese Restaurant  \
0                  Croydon          0.020833                  0.00   
1              Westminster          0.000000                  0.00   
2                Islington          0.000000                  0.00   
3               Wandsworth          0.026667                  0.00   
4           Waltham Forest          0.000000                  0.00   
5            Tower Hamlets          0.000000                  0.00   
6                   Sutton          0.000000                  0.00   
7                Southwark          0.010000                  0.00   
8     Richmond upon Thames          0.000000                  0.00   
9                Redbridge          0.000000                  0.00   
10                  Newham          0.000000                  0.00   
11                  Merton          0.000000                  0.00   
12                Lewisham          0.000000                  0.00   
13                 Lambeth          0.000000                  0.01   
14    Kingston upon Thames          0.010000                  0.00   
15  Kensington and Chelsea          0.010000                  0.00   
16                  Barnet          0.000000                  0.00   
17                  Bexley          0.000000                  0.00   
18              Hillingdon          0.013889                  0.00   
19                Havering          0.000000                  0.00   
20                  Harrow          0.000000                  0.00   
21                Haringey          0.000000                  0.00   
22  Hammersmith and Fulham          0.010000                  0.00   
23                 Hackney          0.000000                  0.00   
24               Greenwich          0.017241                  0.00   
25                 Enfield          0.000000                  0.00   
26                  Ealing          0.010000                  0.00   
27                  Camden          0.010000                  0.00   
28                 Bromley          0.016129                  0.00   
29                   Brent          0.011111                  0.00   
30                Hounslow          0.015152                  0.00   

    Chinese Restaurant  Dim Sum Restaurant  Dumpling Restaurant  \
0             0.010417                0.00                 0.00   
1             0.010000                0.00                 0.00   
2             0.000000                0.00                 0.00   
3             0.000000                0.00                 0.00   
4             0.000000                0.00                 0.00   
5             0.013699                0.00                 0.00   
6             0.000000                0.00                 0.00   
7             0.000000                0.00                 0.00   
8             0.012987                0.00                 0.00   
9             0.000000                0.00                 0.00   
10            0.026316                0.00                 0.00   
11            0.000000                0.00                 0.00   
12            0.025641                0.00                 0.00   
13            0.000000                0.01                 0.01   
14            0.010000                0.00                 0.00   
15            0.020000                0.00                 0.00   
16            0.000000                0.00                 0.00   
17            0.021277                0.00                 0.00   
18            0.013889                0.00                 0.00   
19            0.000000                0.00                 0.00   
20            0.023810                0.00                 0.00   
21            0.044118                0.00                 0.00   
22            0.000000                0.00                 0.00   
23            0.000000                0.00                 0.00   
24            0.017241                0.00                 0.00   
25            0.000000              

This Dataframe shows that only two Boroughs have Korean restaurants, with Croydon having a frequency of 0.0208 and Westminster having a frequency of 0.01.